guardian/fastly-content-fanout
# Create EventBridge

Co-authored-by @twrichards 

## What does this change?

At The Guardian we often report on events that are taking place in real-time, during which circumstances/facts can change very quickly, for example reporting on politics (e.g. general elections) or sports (e.g. Olympics). We also regularly curate fronts to give readers the best coverage of our journalism. In both of these cases it's important to deliver these updates to readers quickly and efficiently, allowing them to stay informed and up-to-date.

Currently our web and app clients poll servers to retrieve updates - this is wasteful in terms of http requests but it can also (depending on the polling frequency) introduce a lag between updates being published and these being available to readers on their devices. Considering this, coupled with effective caching at the CDN, it can be minutes before readers see the latest news.

In #3 we create a Compute@Edge service to allow clients (web and app) to establish long-lived connections for a specific path ("channel"), i.e. what front/article the reader is currently viewing. Establishing the long-lived connection (WebSockets or Server-Sent Events) is only one part of the overall flow. After the connection to a specific channel has been established we can then use Fastly's [Fanout](https://docs.fastly.com/products/fanout) service to broadcast relevant messages to all connected clients:

![compute-at-edge](https://github.com/guardian/fastly-content-fanout/assets/45561419/1c1b7dc0-5306-4564-a67c-c76dd2df049c)

In this PR we configure an instance [EventBridge](https://docs.aws.amazon.com/eventbridge/) in the cms-fronts account to listen for events relating to updates to fronts. When an event has been received we use an [API destination rule](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-api-destinations.html) to send a https request to Fanout. From this request Fanout broadcasts the messages to connected clients (who've subscribed to the pertinent channel).




## How to test

Given https://github.com/guardian/facia-tool/pull/1535 is deployed into CODE then we should expect:
- when an edit to a given front is published
- then an event will be put onto the EventBridge bus

Given an event arrives onto the EventBridge bus then we should expect:
- the API destination rule to be invoked
- the front "path" to be extracted from the event detail and used in the API request to Fanout
- any clients (web, app, terminal) subscribed to the given channel, to receive a message where the body is the time of the edit

It's not necessary to test the flow with the change to facia-tool (although it is a nice end-end test) - we can put messages onto the EventBridge bus directly through the console, we just need to make sure the event detail is correctly formatted, e.g.:
<img width="855" alt="Screenshot 2023-12-20 at 14 47 46" src="https://github.com/guardian/fastly-content-fanout/assets/45561419/f206eb35-6bda-47f2-b562-a7ca010cc50a">
